File: prob211.py
import functools
from math import floor, sqrt
from collections import Counter

maxno = 64000000
factor = lambda n ,factor: n % factor == 0

@functools.lru_cache(maxsize=None)
def sumgeoseries(a, n):
	"""sum of r= 0 to n a^(2r)"""
	return sum(a**(2*r) for r in range(n + 1))
#	return (a**(2*n + 2) - 1) // (a*a - 1)

perfect_squares = set(i*i for i in range(1, floor(sqrt(maxno))))

@functools.lru_cache(maxsize=None)
def sumsqdiv(n):
	for i in range(2, floor(sqrt(n)) + 1):
		power = 0
		while factor(n ,i):
			power += 1
			n //= i
		if power != 0:
			return sumgeoseries(i, power) * sumsqdiv(n)
	return n*n + 1 # Must be prime if reached

print(1 + sum(i for i in range(2, maxno)
	if sumsqdiv(i) in perfect_squares))
#	if sqrt(sumsqdiv(i)).is_integer()))

File: prob211.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#define MAXNO 64000000

static int perfect_sq(double n)
{
	double root = (long int) sqrt(n);
	return n == root * root;
}

long int sumsqdiv[MAXNO];

double sumgeoseries(double a, int n)
{
	/* sum of r= 0 to n a^(2r) */
	return (pow(a, 2*n + 2) - 1) / (a*a - 1);
}

void init_sumsqdiv(void)
{
	long int x,y;
	sumsqdiv[1] = 1;
	for (x = 2; x < MAXNO; x++) {
		int prime = 1;
		for (y = 2; y < sqrt(x); y++) {
			int power = 0;
			long int n = x;
			while (n % y == 0) {
				n /= y;
				power++;
			}
			if (power != 0) {
				sumsqdiv[x] = sumgeoseries(y, power) * sumsqdiv[n];
				prime = 0;
				break;
			}
		}
		if (prime)
			sumsqdiv[x] = x*x + 1;
	}
}

int main(void)
{
	long int total = 0;
	int i;

	init_sumsqdiv();
	for (i = 1; i < MAXNO; i++) {
		if (perfect_sq(sumsqdiv[i]))
			total += i;
	}

	printf("%ld\n", total);
}
